src/components/ColorPicker.tsx

```diff
@@ -5,46 +5,58 @@ import Saturation from "./Saturation";
 
 import styles from "../styles.css";
 import { ColorModel, HSV, ColorPickerBaseProps, AnyColor } from "../types";
-import { equalColorObjects } from "../utils/compare";
 import formatClassName from "../utils/formatClassName";
 
 interface Props<T extends AnyColor> extends Partial<ColorPickerBaseProps<T>> {
   colorModel: ColorModel<T>;
 }
 
+function useEventCallback<T>(event?: (arg: T) => void) {
+  const eventRef = useRef(event);
+
+  useEffect(() => {
+    eventRef.current = event;
+  });
+
+  return useCallback((arg: T) => eventRef.current && eventRef.current(arg), []);
+}
+
 const ColorPicker = <T extends AnyColor>({
   className = "",
   colorModel,
   color = colorModel.defaultColor,
   onChange,
 }: Props<T>) => {
+  // Save onChange callback in the ref for avoiding "useCallback hell"
+  const onChangeRef = useEventCallback<T>(onChange);
+
   // No matter which color model is used (HEX, RGB or HSL),
   // all internal calculations are based on HSV model
   const [hsv, updateHsv] = useState<HSV>(() => colorModel.toHsv(color));
 
   // By using this ref we're able to prevent extra updates
   // and the effects recursion during the color conversion
-  const cache = useRef({ color, hsv });
+  const cache = useRef(color);
 
   // Update local HSV if `color` property value is changed,
   // but only if that's not the same color that we just sent to the parent
   useEffect(() => {
-    if (!colorModel.equal(color, cache.current.color)) {
+    if (!colorModel.equal(color, cache.current)) {
       const newHsv = colorModel.toHsv(color);
-      cache.current = { hsv: newHsv, color };
+      cache.current = color;
       updateHsv(newHsv);
     }
   }, [color, colorModel]);
 
-  // If HSV is changed, convert it to the output format, send to the parent component
+  // Сonvert HSV to the output format, if it is changed send to the parent component
   // and save the new color to the ref to prevent unnecessary updates
   useEffect(() => {
-    if (!equalColorObjects(hsv, cache.current.hsv)) {
```

Comment:
I understand why you dropped `hsv` from `cache` and killed this condition.
This part of my code looks stupid but it makes sense =)

The problem is hidden inside the color conversion algorithms. They aren't 100% "accurate" for conditions we need here.
Mostly because of `Math.round` and `Math.floor` they have. Sometimes because of the conditions inside the code.

Just try to set `#c62182` as the default color in the demo and you'll see that in your branch `ColorPicker` triggers `onChange` right after the mounting and prints `#c72282` to the console. These colors are almost the same but they are not because during the `hex→rgb→hsv→rgb→hex` we run `Math.floor` and `Math.round` a lot of times and we are not able to get the source color. It's sad but that's how color conversion algorithms work.

```
const someHsv =  hexToHsv('#c62182')
hsvToHex('#c62182') // #c72282' - different color
```

So I store previous HSV in the ref to be able to compare them. Don't like this thing, but it works.